modules/nextflow/src/main/groovy/nextflow/extension/PublishOp.groovy

```diff
@@ -149,38 +201,46 @@ class PublishOp {
      * work directory paths to publish paths.
      *
      * @param value
+     * @param targetDirOrClosure
      */
-    protected Object normalizePaths(value) {
+    protected Object normalizePaths(value, targetDirOrClosure) {
         if( value instanceof Path ) {
-            return List.of(value.getBaseName(), normalizePath(value))
+            return List.of(value.getBaseName(), normalizePath(value, targetDirOrClosure))
         }
 
         if( value instanceof Collection ) {
             return value.collect { el ->
                 if( el instanceof Path )
-                    return normalizePath(el)
+                    return normalizePath(el, targetDirOrClosure)
                 if( el instanceof Collection<Path> )
-                    return normalizePaths(el)
+                    return normalizePaths(el, targetDirOrClosure)
                 return el
             }
         }
 
         if( value instanceof Map ) {
-            return value.collectEntries { k, v ->
-                if( v instanceof Path )
-                    return List.of(k, normalizePath(v))
-                if( v instanceof Collection<Path> )
-                    return List.of(k, normalizePaths(v))
-                return List.of(k, v)
-            }
+            return value
+                .findAll { k, v -> v != null }
+                .collectEntries { k, v ->
+                    if( v instanceof Path )
+                        return List.of(k, normalizePath(v, targetDirOrClosure))
+                    if( v instanceof Collection<Path> )
+                        return List.of(k, normalizePaths(v, targetDirOrClosure))
+                    return List.of(k, v)
+                }
         }
 
         throw new IllegalArgumentException("Index file record must be a list, map, or file: ${value} [${value.class.simpleName}]")
     }
 
-    private Path normalizePath(Path path) {
+    private String normalizePath(Path path, targetDirOrClosure) {
+        if( targetDirOrClosure instanceof Closure )
+            return (targetDirOrClosure.call(path.getName()) as Path).normalize().toUriString()
```

Comment:
It would always be better to use `FileHelper.asPath` over `as Path`. Even if it's doing be same under the hood, it would allow us to better keep where this conversion is done.

Reply:
Agreed